File: windows/src/ControlDeck.Domain/GroupController.cs
namespace ControlDeck.Domain;

/// <summary>
/// One outbound command produced by group math: which device to send it to,
/// which widget it logically corresponds to (for UI feedback/correlation),
/// and the resolved absolute/edge action.
/// </summary>
public sealed record GroupDispatch(DeviceId TargetDeviceId, WidgetId WidgetId, ActionSpec Action);

/// <summary>A RELATIVE_SLIDER group member's current absolute value and the widget's own action template.</summary>
public sealed record RelativeSliderMember(WidgetId WidgetId, DeviceId TargetDeviceId, int CurrentValue, ActionSpec ActionTemplate);

/// <summary>An ABSOLUTE_TOGGLE (mute) group member's current mute state.</summary>
public sealed record MuteMember(WidgetId WidgetId, DeviceId TargetDeviceId, bool CurrentlyMuted);

/// <summary>An ABSOLUTE_MEDIA group member's current playback state.</summary>
public sealed record MediaMember(WidgetId WidgetId, DeviceId TargetDeviceId, MediaState CurrentState);

/// <summary>
/// Pure group math, centralized per docs/ARCHITECTURE.md §4. UI layers call
/// these functions and never compute deltas or target states themselves —
/// no I/O happens here; callers are responsible for supplying current state
/// and for actually sending the resulting <see cref="GroupDispatch"/>es over
/// the transport.
/// </summary>
public static class GroupController
{
    /// <summary>
    /// §4.1 RELATIVE_SLIDER: delta is applied to every member's current value,
    /// then each member is independently clamped to [0,100] — no redistribution
    /// of the delta across the group. A member whose clamped value doesn't
    /// change (already saturated in the direction of travel) is skipped so no
    /// redundant no-op ACTION is sent.
    /// </summary>
    public static IReadOnlyList<GroupDispatch> ApplyRelativeSlider(
        IReadOnlyList<RelativeSliderMember> members,
        int oldValue,
        int newValue)
    {
        var delta = newValue - oldValue;
        var dispatches = new List<GroupDispatch>();

        foreach (var member in members)
        {
            var clamped = Math.Clamp(member.CurrentValue + delta, 0, 100);
            if (clamped == member.CurrentValue)
            {
                continue;
            }

            ActionSpec action = member.ActionTemplate switch
            {
                BrightnessSet => new BrightnessSet(clamped),
                VolumeSet => new VolumeSet(clamped),
                _ => throw new InvalidOperationException(
                    $"Relative slider group member {member.WidgetId} has an action template " +
                    $"({member.ActionTemplate.GetType().Name}) that is not a value-bearing slider action.")
            };

            dispatches.Add(new GroupDispatch(member.TargetDeviceId, member.WidgetId, action));
        }

        return dispatches;
    }

    /// <summary>
    /// §4.2 ABSOLUTE_TOGGLE (mute groups): mute everything unless everything is
    /// already muted, in which case unmute everything. A member's local state
    /// is never inverted independently of the others — the desired absolute
    /// state is computed once and broadcast to every member.
    /// </summary>
    public static IReadOnlyList<GroupDispatch> ApplyAbsoluteToggle(IReadOnlyList<MuteMember> members)
    {
        if (members.Count == 0)
        {
            return Array.Empty<GroupDispatch>();
        }

        var allMuted = members.All(m => m.CurrentlyMuted);
        var desiredMuted = !allMuted;

        return members
            .Select(m => new GroupDispatch(m.TargetDeviceId, m.WidgetId, new SetMuted(desiredMuted)))
            .ToList();
    }

    /// <summary>
    /// §4.3 ABSOLUTE_MEDIA, explicit variant: the user picked a desired state
    /// directly via a play/pause button pair. Broadcast unconditionally to
    /// every member — no ambiguity to resolve.
    /// </summary>
    public static IReadOnlyList<GroupDispatch> ApplyAbsoluteMedia(
        IReadOnlyList<MediaMember> members,
        MediaState desiredState)
    {
        return members
            .Select(m => new GroupDispatch(m.TargetDeviceId, m.WidgetId, new MediaSetState(desiredState)))
            .ToList();
    }

    /// <summary>
    /// §4.3 ABSOLUTE_MEDIA, single-button toggle variant: when the UI exposes
    /// only one play/pause button (no explicit pair), the desired state is the
    /// opposite of the majority current state across members. Ties (equal
    /// playing/paused counts) resolve to PAUSED, i.e. "not unanimously playing"
    /// is treated the same as "majority paused".
    /// </summary>
    public static IReadOnlyList<GroupDispatch> ApplyAbsoluteMediaToggle(IReadOnlyList<MediaMember> members)
    {
        var desiredState = MajorityState(members) == MediaState.Playing ? MediaState.Paused : MediaState.Playing;
        return ApplyAbsoluteMedia(members, desiredState);
    }

    /// <summary>
    /// §4.3: MEDIA_NEXT/MEDIA_PREVIOUS are edge-triggered and always
    /// absolute-broadcast to every member unconditionally — there is no
    /// "state" to reconcile for a skip event.
    /// </summary>
    public static IReadOnlyList<GroupDispatch> ApplyMediaEdge(IReadOnlyList<MediaMember> members, bool isNext)
    {
        ActionSpec action = isNext ? new MediaNext() : new MediaPrevious();
        return members
            .Select(m => new GroupDispatch(m.TargetDeviceId, m.WidgetId, action))
            .ToList();
    }

    private static MediaState MajorityState(IReadOnlyList<MediaMember> members)
    {
        if (members.Count == 0)
        {
            return MediaState.Paused;
        }

        var playingCount = members.Count(m => m.CurrentState == MediaState.Playing);
        var pausedCount = members.Count - playingCount;
        return playingCount > pausedCount ? MediaState.Playing : MediaState.Paused;
    }
}

File: android/domain/src/main/kotlin/com/controlldeck/domain/GroupController.kt
package com.controlldeck.domain

/**
 * One outbound command produced by group math: which widget triggered it
 * (for UI bookkeeping), which device it must be sent to, and the resulting
 * [ActionSpec]. [GroupController] never performs I/O itself — callers
 * (Action Engine) are responsible for actually sending these over the
 * transport. This is what makes the group math unit-testable in isolation.
 */
data class GroupDispatch(
    val widgetId: WidgetId,
    val targetDeviceId: DeviceId,
    val action: ActionSpec,
)

/** User-originated interactions that can trigger group semantics. */
sealed class GroupUserInput {
    /** A relative slider (brightness/volume) was dragged from [oldValue] to [newValue]. */
    data class SliderMoved(val originWidgetId: WidgetId, val oldValue: Int, val newValue: Int) : GroupUserInput()

    /** The single mute/toggle button for an ABSOLUTE_TOGGLE group was pressed. */
    data class ToggleActivated(val originWidgetId: WidgetId) : GroupUserInput()

    /**
     * A play/pause **button pair** was pressed with an explicit desired
     * state (docs/ARCHITECTURE.md §4.3, preferred form — unambiguous).
     */
    data class MediaDesiredState(val originWidgetId: WidgetId, val desired: MediaPlaybackState) : GroupUserInput()

    /**
     * A single play/pause toggle button was pressed (ambiguous form): the
     * desired state is derived from the majority of current member states.
     */
    data class MediaToggle(val originWidgetId: WidgetId) : GroupUserInput()

    /** MEDIA_NEXT/MEDIA_PREVIOUS are edge-triggered and always broadcast unconditionally. */
    data class MediaEdge(val originWidgetId: WidgetId, val action: ActionSpec) : GroupUserInput()
}

/**
 * Snapshot of everything [GroupController] needs to compute group math,
 * supplied by the caller (Action Engine / ViewModel) — never fetched by
 * this class, keeping it pure and side-effect free.
 */
class GroupContext(
    private val widgetsById: Map<WidgetId, Widget>,
    private val deviceStates: Map<DeviceId, DeviceState>,
) {
    fun widget(id: WidgetId): Widget? = widgetsById[id]

    fun currentValue(widget: Widget): Int? = when (widget.type) {
        WidgetType.SLIDER_BRIGHTNESS -> deviceStates[widget.targetDeviceId]?.brightness
        WidgetType.SLIDER_VOLUME -> deviceStates[widget.targetDeviceId]?.volume
        else -> null
    }

    fun currentMuted(widget: Widget): Boolean? = deviceStates[widget.targetDeviceId]?.muted

    fun currentMediaState(widget: Widget): MediaPlaybackState? = deviceStates[widget.targetDeviceId]?.mediaState

    companion object {
        fun of(widgets: List<Widget>, deviceStates: List<DeviceState>): GroupContext =
            GroupContext(widgets.associateBy { it.id }, deviceStates.associateBy { it.deviceId })
    }
}

/**
 * Pure implementation of the three group algorithms in
 * docs/ARCHITECTURE.md §4. UI layers must always route group interactions
 * through [apply] rather than computing deltas/targets themselves, so
 * Android and Windows controllers cannot drift apart.
 */
object GroupController {

    fun apply(group: Group, context: GroupContext, input: GroupUserInput): List<GroupDispatch> {
        val members = group.memberWidgetIds.mapNotNull { context.widget(it) }
        return when (group.kind) {
            GroupKind.RELATIVE_SLIDER -> applyRelativeSlider(members, context, input)
            GroupKind.ABSOLUTE_TOGGLE -> applyAbsoluteToggle(members, context, input)
            GroupKind.ABSOLUTE_MEDIA -> applyAbsoluteMedia(members, context, input)
        }
    }

    /** ARCHITECTURE.md §4.1 — independent clamp(0,100) per member, no redistribution. */
    private fun applyRelativeSlider(
        members: List<Widget>,
        context: GroupContext,
        input: GroupUserInput,
    ): List<GroupDispatch> {
        val moved = input as? GroupUserInput.SliderMoved ?: return emptyList()
        val delta = moved.newValue - moved.oldValue
        if (delta == 0 || members.isEmpty()) return emptyList()

        val dispatches = mutableListOf<GroupDispatch>()
        for (member in members) {
            val current = context.currentValue(member) ?: continue
            val proposed = clampPercent(current + delta)
            // Skip members already saturated in the direction of travel: no-op traffic avoidance.
            if (proposed == current) continue
            val action = when (member.type) {
                WidgetType.SLIDER_BRIGHTNESS -> ActionSpec.BrightnessSet(proposed)
                WidgetType.SLIDER_VOLUME -> ActionSpec.VolumeSet(proposed)
                else -> continue
            }
            dispatches += GroupDispatch(member.id, member.targetDeviceId, action)
        }
        return dispatches
    }

    /** ARCHITECTURE.md §4.2 — mute everything unless everything is already muted. */
    private fun applyAbsoluteToggle(
        members: List<Widget>,
        context: GroupContext,
        input: GroupUserInput,
    ): List<GroupDispatch> {
        if (input !is GroupUserInput.ToggleActivated) return emptyList()
        if (members.isEmpty()) return emptyList()

        val allMuted = members.all { context.currentMuted(it) == true }
        val desiredMuted = !allMuted
        return members.map { member ->
            GroupDispatch(member.id, member.targetDeviceId, ActionSpec.SetMuted(desiredMuted))
        }
    }

    /** ARCHITECTURE.md §4.3 — explicit desired state, majority-derived toggle, or edge-triggered skip. */
    private fun applyAbsoluteMedia(
        members: List<Widget>,
        context: GroupContext,
        input: GroupUserInput,
    ): List<GroupDispatch> {
        if (members.isEmpty()) return emptyList()

        return when (input) {
            is GroupUserInput.MediaDesiredState -> members.map { member ->
                GroupDispatch(member.id, member.targetDeviceId, ActionSpec.MediaSetState(input.desired))
            }

            is GroupUserInput.MediaToggle -> {
                val states = members.mapNotNull { context.currentMediaState(it) }
                val playingCount = states.count { it == MediaPlaybackState.PLAYING }
                val pausedCount = states.count { it == MediaPlaybackState.PAUSED }
                val majorityIsPlaying = playingCount >= pausedCount && playingCount > 0
                val desired = if (majorityIsPlaying) MediaPlaybackState.PAUSED else MediaPlaybackState.PLAYING
                members.map { member ->
                    GroupDispatch(member.id, member.targetDeviceId, ActionSpec.MediaSetState(desired))
                }
            }

            is GroupUserInput.MediaEdge -> members.map { member ->
                GroupDispatch(member.id, member.targetDeviceId, input.action)
            }

            else -> emptyList()
        }
    }
}
